new rule: may use asynchronous processing (#787)
zalando/restful-api-guidelines: chapters/http-requests.adoc

would be a similar well defined option.

[#253]
== {MAY} use asynchronous request processing

Typically REST APIs are designed as synchronous interfaces where all server-side 
processing and state changes initiated by the call are finished before delivering the 
result as request response. However, in long running request processing situations 
you may make use of asynchronous interface design with multiple calls: one for 
initiating the asynchronous processing and subsequent ones for accessing the 
processing status and/or result.

We recommend an API design that represents the asynchronous request processing 
explicitly via a job resource that has a status and is different from the actual 

> business resource. For instance, `POST /report-job` returns http status code {202} to 

Reviewer: Shouldn't this be 200? After all, creating a report is asynchronous, but creating a report job is synchronous.

Author: yes, typo -- will change to 200

indicate successful initiation of asynchronous processing together with the job-id 

> as response payload. The job-id then can be used to poll the processing status 

Author: as discussed...
a) 201 should be returned here
b) job-id can be passed via payload and/or location return header
c) location header (instead of 'artificial' report payload) should be returned in the POST /report situation below
d) let's discuss in the API Guild whether we should introduce a new rule requiring to always sent the location header additionally to the payload

> as response payload. The job-id then can be used to poll the processing status 

Author: @tkrop implemented the changes mentioned above and put topic d) on the API Guild agenda. Pls. recheck / approve, thx.

> via `GET /report-job/{id}` which returns http status code {200} with job status and 

Reviewer: Resource should be plural, right? Also in other places.

```suggestion
via `GET /report-jobs/{id}` which returns http status code {200} with job status and 
```

> optional report-id as response payload. Once returned, the report-id can be used 

Reviewer: Why is report-id optional?

Author: The ideas is that the report-id is only delivered, if the processing is in finished state and the identified report exists.

Author: added a hint to make it clearer

to fetch the report result via `GET /report/{id}` which returns http status code {200}  
and the report object as response payload. 

In case you do not follow the best practice via a separate job resource, `POST /report`
returns http status code {202} to indicate successful initiation of asynchronous 
processing together with the report-id as response payload. The report-id can be 
used to fetch the report result via `GET /report/{id}` which returns either http status 
code {200} and the report object response payload or status code {202} without the 
report object response payload depending on the asynchronous processing status.

> Hint: Do _not_ use response code {204} or {404} instead {202} here -- it is missleading

Reviewer: ```suggestion
Hint: Do _not_ use response code {204} or {404} instead {202} here -- it is misleading
```

would be a similar well defined option.

[#253]
== {MAY} support asynchronous request processing

Typically REST APIs are designed as synchronous interfaces where all server-side 
processing and state changes initiated by the call are finished before delivering
the result as response. However, in long running request processing situations 
you may make use of asynchronous interface design with multiple calls: one for 
initiating the asynchronous processing and subsequent ones for accessing the 
processing status and/or result.

We recommend an API design that represents the asynchronous request processing 
explicitly via a job resource that has a status and is different from the actual 

> business resource. For instance, `POST /report-jobs` returns http status code {201} to 

Reviewer 2: Either
```suggestion
business resource. For instance, `POST /report-jobs` returns HTTP status code {201} to 
```
or shorter
```suggestion
business resource. For instance, `POST /report-jobs` returns a status code {201} to 
```
or very short
```suggestion
business resource. For instance, `POST /report-jobs` returns {201} to 
```

indicate successful initiation of asynchronous processing together with the job-id 

> passed with the response payload and/or {Location} header. 

Reviewer 2: ```suggestion
passed in the response payload and/or via the URL of the {Location} header. 
```

The job-id or {Location} URL then can be used to poll the processing status 

> via `GET /report-jobs/{id}` which returns http status code {200} with job status and 

Reviewer 2: Either
```suggestion
via `GET /report-jobs/{id}` which returns HTTP status code {200} with job status and 
```
or
```suggestion
via `GET /report-jobs/{id}` which returns a status code {200} with job status and 
```
or very short
```suggestion
via `GET /report-jobs/{id}` which returns {200} with job status and 
```
